packages/client/src/devices/InputMediaDeviceManager.ts

```diff
@@ -142,20 +151,29 @@ export abstract class InputMediaDeviceManager<
    * @param filter the filter to register.
    * @returns a function that will unregister the filter.
    */
-  async registerFilter(
-    filter: MediaStreamFilter,
-    cleanup?: MediaStreamFilterCleanup,
-  ) {
+  registerFilter(filter: MediaStreamFilter) {
     const entry: [MediaStreamFilter, MediaStreamFilterCleanup] = [
       filter,
-      cleanup,
+      undefined,
     ];
-    this.filters.push(entry);
-    await this.applySettingsToStream();
-    return async () => {
-      cleanup?.();
-      this.filters = this.filters.filter((f) => f !== entry);
-      await this.applySettingsToStream();
+
+    const ready = withoutConcurrency(
+      this.filterRegistrationConcurrencyTag,
+      async () => {
+        this.filters.push(entry);
+        await this.applySettingsToStream();
+      },
+    );
+
+    return {
+      ready,
+      unregister: () =>
+        withoutConcurrency(this.filterRegistrationConcurrencyTag, async () => {
+          const cleanup = entry[1];
+          cleanup?.();
+          this.filters = this.filters.filter((f) => f !== entry);
+          await this.applySettingsToStream();
+        }),
     };
```

Comment:
We move the logic preventing unregistering before finishing registering into the client. Note how both registering and unregistering logic is wrapped in `withoutConcurrency`.

This can be further optimized. Most of the times we unregister one filter to immediately register another. We can introduce a method called `replaceFilter` that will only call `applySettingsToStream` once. That way we won't get flickering when switching blur levels or background images.

Reply:
I don't like the API with filters returning [stream, cleanup] tuple, leaving it in this draft, but will probably change before merging.